### DISCUSSION TITLE
Isomorphic fetch doesnt set cookies

### DISCUSSION THREAD

**Discussion by mihail727** [nuxt/nuxt#17687] [Questions] [answered]
### Environment

------------------------------
- Operating System: `Linux`
- Node Version:     `v16.14.0`
- Nuxt Version:     `3.0.0-rc.0-27508091.78fcbcf`
- Package Manager:  `yarn@1.22.18`
- Builder:          `vite`
- User Config:      `runtimeConfig`, `buildModules`, `modules`, `tailwindcss`, `app`, `nitro`, `build`, `modern`
- Runtime Modules:  `@nuxtjs/tailwindcss@5.0.3`
- Build Modules:    `@vueuse/nuxt@8.3.1`
------------------------------

### Reproduction

*

### Describe the bug

I want to make my own universal fetch, which would automatically send the token stored in the client's cookies
And if the token is invalid, it was automatically updated
What i have:
`@/composables/useRequest.ts`
```ts
export async function useRequest<T>(
	url: string,
	options: { method?: 'GET' | 'POST'; body?: {}; headers?: HeadersInit } = {
		body: null,
		method: 'GET',
	},
) {
	return await $fetch(url, options).catch(async (e) => {
		if (!e?.message?.includes('401')) throw e;

		const response = await $fetch.raw<IServerResponse<string>>('/api/auth/refresh', {
			headers: options.headers,
		});

		if (process.server) {
			const cookies = Object.fromEntries(
				response.headers
					.get('set-cookie')
					?.split(',')
					.map((a) => a.split('=')),
			);

			const cookieName = 'access';

			if (cookieName in cookies) {
				const newCookie = cookies[cookieName]?.split(';')[0];
				useCookie(cookieName, { maxAge: 7 * 24 * 60 * 60 }).value = newCookie;
			}
		}
		// return await $fetch<T>(url, options);
	});
}
```
on the server side in apiroutes i set new cookies like this:
`setCookie(event, 'access', tokens.accessToken, { maxAge: 7 * 24 * 60 * 60 });`

the most incomprehensible thing for me - is in this line:
`@/composables/useRequest.ts`
```ts
if (cookieName in cookies) {
	console.log('\n useRequest SERVER_SIDE COOKIES :::', cookies[cookieName]);
	const newCookie = cookies[cookieName]?.split(';')[0];
	useCookie(cookieName, { maxAge: 7 * 24 * 60 * 60 }).value = newCookie;
	console.log('\n useRequest SERVER_SIDE useCookie :::', useCookie(cookieName).value); // [THIS ROW]
}
```
[THIS ROW] never be shown in my server console...
and ofc i dont have new cookies at my client side

**Comment by danielroe** [comment, marked as answer]
All of your handler is only running when you receive a fetch error (it's in a catch callback).

**Reply by mihail727** [reply]
wait..... what you mean
I understand yes, 401 status code gives middleware
as soon as i get 401 i try to refresh the token.
all the code inside the catch works except for what I described

**Reply by mihail727** [reply]
`@/server/middleware/access-middleware.ts`
```ts
import { useCookie } from 'h3';
import { tokenService } from '../services';

export default defineEventHandler((event) => {
	const included_urls = ['/api/test'];
	const bypass = included_urls.some((url) => event.req.url.includes(url));
	if (bypass) {
		const access_token = useCookie(event, 'access');
		if (!tokenService.verifyToken(access_token)) {
			console.log('\nACCESS MIDDLEWARE INVALID TOKEN\n', access_token);
			event.res.statusCode = 401;
			return { success: false, reason: 'Invalid access token' };
		}
	}
});

```

**Reply by mihail727** [reply]
little reproduction
https://stackblitz.com/edit/nuxt-starter-fhtdqk?file=server/api/refresh.get.ts

**Reply by mihail727** [reply]
in reproduction it should always set new cookies

**Reply by danielroe** [reply]
There's an issue that you're not seeing an error message, which we need to look at. But, in brief, you can't access the Nuxt instance after an async call. (We add special handling in plugins, middleware & setup to preserve the nuxt instance.)

So you should rewrite your composable like this:

```js
export async function useRequest<T>(
  url: string,
  options: { method?: 'GET' | 'POST'; body?: {}; headers?: HeadersInit } = {
    body: null,
    method: 'GET',
  }
) {
  const accessCookie = useCookie('access', { maxAge: 7 * 24 * 60 * 60 })
  try {
    return await $fetch(url, options)
  } catch (e) {
    if (!e?.message?.includes('401')) throw e

    const response = await $fetch.raw('/api/refresh', {
      headers: options.headers,
    })

    if (process.server) {
      const cookies = Object.fromEntries(
        response.headers
          .get('set-cookie')
          ?.split(',')
          .map(a => a.split('='))
      )

      if ('access' in cookies) {
        accessCookie.value = cookies.access?.split(';')[0]
      }
    }
  }
}
```

I would further observe that composable (with use) should in general not be async, but should call whatever setup functions (like useCookie) it needs, and return _handlers_ that can be async.

**Reply by mihail727** [reply]
omg big thanks..... i suffered from this problem for about a month

**Reply by mihail727** [reply]
here my final solution: https://github.com/nuxt/framework/discussions/4597
